x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvMedianAbsoluteDeviation.java

```diff
@@ -0,0 +1,335 @@
+
+    private MvMedianAbsoluteDeviation(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected TypeResolution resolveFieldType() {
+        return isType(field(), t -> t.isNumeric() && isRepresentable(t), sourceText(), null, "numeric");
+    }
+
+    @Override
+    protected ExpressionEvaluator.Factory evaluator(ExpressionEvaluator.Factory fieldEval) {
+        return switch (PlannerUtils.toElementType(field().dataType())) {
+            case DOUBLE -> new MvMedianAbsoluteDeviationDoubleEvaluator.Factory(fieldEval);
+            case INT -> new MvMedianAbsoluteDeviationIntEvaluator.Factory(fieldEval);
+            case LONG -> field().dataType() == DataType.UNSIGNED_LONG
+                ? new MvMedianAbsoluteDeviationUnsignedLongEvaluator.Factory(fieldEval)
+                : new MvMedianAbsoluteDeviationLongEvaluator.Factory(fieldEval);
+            default -> throw EsqlIllegalArgumentException.illegalDataType(field.dataType());
+        };
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new MvMedianAbsoluteDeviation(source(), newChildren.get(0));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, MvMedianAbsoluteDeviation::new, field());
+    }
+
+    static class Doubles {
+        public double[] values = new double[2];
+        public int count;
+    }
+
+    @MvEvaluator(extraName = "Double", finish = "finish", single = "single")
+    static void process(Doubles doubles, double v) {
+        if (doubles.values.length < doubles.count + 1) {
+            doubles.values = ArrayUtil.grow(doubles.values, doubles.count + 1);
+        }
+        doubles.values[doubles.count++] = v;
+    }
+
+    static double finish(Doubles doubles) {
+        double median = doubleMedianOf(doubles.values, doubles.count);
+        for (int i = 0; i < doubles.count; i++) {
+            double value = doubles.values[i];
+            doubles.values[i] = value > median ? value - median : median - value;
+            assert Double.isFinite(doubles.values[i]) : "Overflow on median differences";
+        }
+        double mad = doubleMedianOf(doubles.values, doubles.count);
+        doubles.count = 0;
+        return mad;
+    }
+
+    static double ascending(Doubles doubles, DoubleBlock values, int firstValue, int count) {
+        if (doubles.values.length < count) {
+            doubles.values = ArrayUtil.grow(doubles.values, count);
+        }
+        int middle = firstValue + count / 2;
+        double median = count % 2 == 1 ? values.getDouble(middle) : (values.getDouble(middle - 1) + values.getDouble(middle)) / 2;
+        for (int i = 0; i < count; i++) {
+            double value = values.getDouble(firstValue + i);
+            doubles.values[i] = value > median ? value - median : median - value;
```

Comment:
This currently overflows, as the abs difference between the median and a big/smalll number may be higher than MAX_TYPE.
A "simple" fix would be to change:
- For longs, use BigInteger
- For doubles, use BigDecimal
- For ints, use longs
But using those for every calculation would introduce a lot of overhead (Specially for doubles and longs).

Other options I considered:
- Throw and return null. This is not a median, and this may overflow. The cases are very... Edgy, so could be fine. No overflow, no overhead.
- Detect overflows in a per-element basis, and only throw if one of those overflows has to be used. Quite the logic, and I don't think it's possible to have a MAD higher than MAX_VALUE. Which leads me to the next option:
- Replace over/underflows with MAX/MIN values, just to sort them, and later ignore them? If the MAD can't be higher than MAX_VALUE, then this should render a correct sorting and a correct MAD

Dumping my ideas here, so others can check and validate them. The last one feels nice to me, but I still have to do some checks to avoid committing a mathematical crime here

Reply:
Personally, I'm in favor of collecting the overflowing values and return `null` + a warning if and only if we determine that the median absolute deviation has to be properly out of range.

From the top of my head, I cannot determine an edge case where the MAD itself would be a legitimate overflow - maybe with longs we can construct such cases (because the negative and positive max values are asymmetric), but with doubles it may be properly impossible (in theory, not accounting for rounding errors). If that's correct, trying to simply keep track of the cases when the distance to the median is out of range _should_ be sufficient.

Reply:
After some thought I'm following a mixed approach:
- For ints, _just_ using longs instead
- For longs, calculating the differences as unsigned longs. Looks like the more strict and realistic approach to me. After all, differences are, indeed, unsigned
- For doubles, I'll just let it infinite. As Infinities are sortable, they should work well

Both ints and longs are checked on down-conversion to make sure they don't overflow

Reply:
Sounds very convincing!

Maybe it's worth asserting that the result is never overflowing, resp. never an infinity; I agree that we shouldn't proactively null+warn if we never expect this to be relevant.